titan/model.py

```diff
@@ -358,48 +362,108 @@ def agents_interact(self, rel: "ag.Relationship"):
             interaction = self.interactions[interaction_type]
             interaction.interact(self, rel)
 
-    def die_and_replace(self):
+    def exit(self):
         """
-        Let agents die and replace the dead agent with a new agent randomly.
+        Allow agents to exit model.
         """
-        # die stage
-        for agent in self.pop.all_agents:
-            # agent incarcerated, don't evaluate for death
-            if agent.incar.active:
-                continue
-
-            # death rate per 1 person-month
-            p = (
-                prob.get_death_rate(
-                    agent.hiv.active,
-                    agent.hiv.aids,
-                    agent.drug_type,
-                    agent.sex_type,
-                    agent.haart.adherent,
-                    agent.race,
-                    agent.location,
-                    self.params.model.time.steps_per_year,
-                )
-                * self.calibration.mortality
-            )
-
-            if self.run_random.random() < p:
-                self.deaths.append(agent)
-
-                # End all existing relationships
-                for rel in copy(agent.relationships):
-                    rel.progress(force=True)
-                    self.pop.remove_relationship(rel)
-
-        # replace stage
-        for agent in self.deaths:
-            # mark agent component as -1 (no componenet)
-            agent.component = "-1"
-
-            # Remove agent from agent class and sub-sets
-            self.pop.remove_agent(agent)
+        if self.exits == {}:
```

Comment:
do you think it's worth short-circuiting if there are only `none` exit types as well? Or maybe that's an argument for reversing the loop order so that on the none strategies we don't loop over all the agents?

Reply:
Yes, that's a good idea!